pkg/remediator/watch/listwatch.go

```diff
@@ -0,0 +1,338 @@
+			}
+			// all namespaces
+			return dynamicClient.Resource(mapping.Resource).List(ctx, options)
+		default:
+			return nil, fmt.Errorf("invalid resource scope %q for resource %q", mapping.Scope, mapping.Resource)
+		}
+	}
+	watchFunc := func(ctx context.Context, options metav1.ListOptions) (watch.Interface, error) {
+		options.Watch = true
+		optionsModifier(&options)
+		mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		if err != nil {
+			return nil, errors.Wrapf(err, "watcher failed to get REST mapping for %s", gvk.String())
+		}
+		switch mapping.Scope.Name() {
+		case meta.RESTScopeNameRoot:
+			if namespace != "" {
+				return nil, fmt.Errorf("cannot watch cluster-scoped resource %q in namespace %q", mapping.Resource, namespace)
+			}
+			// cluster-scope
+			return dynamicClient.Resource(mapping.Resource).Watch(ctx, options)
+		case meta.RESTScopeNameNamespace:
+			if namespace != "" {
+				return dynamicClient.Resource(mapping.Resource).Namespace(namespace).Watch(ctx, options)
+			}
+			// all namespaces
+			return dynamicClient.Resource(mapping.Resource).Watch(ctx, options)
+		default:
+			return nil, fmt.Errorf("invalid resource scope %q for resource %q", mapping.Scope, mapping.Resource)
+		}
+	}
+	return &ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
+}
+
+// List a set of apiserver resources
+func (lw *ListWatch) List(ctx context.Context, options metav1.ListOptions) (*unstructured.UnstructuredList, error) {
+	// ListWatch is used in Reflector, which already supports pagination.
+	// Don't paginate here to avoid duplication.
+	return lw.ListFunc(ctx, options)
+}
+
+// Watch a set of apiserver resources
+func (lw *ListWatch) Watch(ctx context.Context, options metav1.ListOptions) (watch.Interface, error) {
+	return lw.WatchFunc(ctx, options)
+}
+
+type listAndWatch struct {
+	stopOnce sync.Once
+	stopCh   chan struct{}
+	eventCh  chan watch.Event
+}
+
+func (ww *listAndWatch) Stop() {
+	ww.stopOnce.Do(func() {
+		close(ww.stopCh)
+	})
+	<-ww.eventCh
+}
+
+func (ww *listAndWatch) ResultChan() <-chan watch.Event {
+	return ww.eventCh
+}
+
+// ListAndWatch wraps a list and watch and returns a watch interface.
+// This way, you can watch from "now" and get all the existing objects, not just
+// new changes to objects.
+//
+// If ResourceVersion is specified, ListAndWatch acts like a normal Watch.
+// Otherwise, a List is performed first and `Added` events are simulated,
+// followed by a Watch, where subsequent `Added` events for pre-existing objects
+// are converted to `Modified` events, unless that object is deleted first.
+func ListAndWatch(ctx context.Context, lw ListerWatcher, opts metav1.ListOptions) (watch.Interface, error) {
+	if opts.ResourceVersion != "" {
+		// Just watch!
+		return lw.Watch(ctx, opts)
+	}
+
+	opts.AllowWatchBookmarks = false // not allowed for List
+	opts.Watch = false               // not allowed for List
+
+	// TODO: Use a short timeout for List and longer for Watch
```

Comment:
What's stopping this from happening now?

Reply:
Wasn't sure how to configure it cleanly.
I could hard code it, but that seemed wrong.
Also seems wrong to pass two timeouts into ListAndWatch.
I could do in the ListerWatcherFactory, but then I have to pass it through watcherConfig and duplicate the res.Config, because it only has one timeout setting, and I don't want to pass two rest.Config through a bunch of different layers when they're mostly identical.
Ideally rest.Config would be modified to support it, but that's an upstream client-go abstraction...

Reply:
I think I can add a wrapper that returns different configs for Get, List, and Watch, but I'll need to do it in another PR, since it will touch everything...

Reply:
Ack, that seems reasonable

Reply:
I spent hours on this and it's a giant headache.

There's a half dozen places where I'm not sure whether it should get the longer timeout or the shorter timeout, because the same client is used for CRUD, List, and Watch...

Today, only the remediator uses the longer watch timeout. Everything else uses the shorter timout, including the applier, which does its own watches, and the discovery clients, which poll long lists of CRDs...